# Ejercicio 51

Obtener las primeras `n` filas de un objeto `DataFrame`.

## Solución:

In [1]:
import pandas as pd
import numpy as np

In [2]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [12]:
print(df.iloc[:3,:]) # la coma separa la notacion desde hasta de flas y columnas, dentro de las filas digo desde a " " : hasta la 3
#print(df.head(3))


    nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí


In [ ]:
df.iloc[0:3]


# Ejercicio 52

Obtener las últimas `n` filas de un objeto `DataFrame`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

In [ ]:
df.iloc[-5:]

# Ejercicio 53

Obtener la representación en **NumPy** de un objeto `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [13]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [14]:
df.to_numpy()

array([[ 1,  5,  2],
       [ 2,  4,  3],
       [ 3,  3,  5],
       [ 4,  2,  7],
       [ 5,  1, 11]], dtype=int64)

# Ejercicio 54

Seleccionar columnas de un objeto `DataFrame` usando notación de slicing.

## Solución:

In [15]:
import pandas as pd
import numpy as np

In [16]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [17]:
df[['nombre', 'califico']]

nombre califico
a    Oliva       Sí
b  Daniela       No
c     Juan       Sí
d   Germán       No
e   Edward       No
f     Alex       Sí
g    Julio       Sí
h    Edgar       No
i    Angie       No
j   Irlesa       Sí

# Ejercicio 55

Obtener los ejes (atributos de filas y columnas) de un `DataFrame`.

## Solución:

In [18]:
import pandas as pd

In [19]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [20]:
df.axes

[RangeIndex(start=0, stop=5, step=1), Index(['X', 'Y', 'Z'], dtype='object')]

In [21]:
indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [22]:
df.axes

[Index(['a', 'b', 'c', 'd', 'e'], dtype='object'),
 Index(['X', 'Y', 'Z'], dtype='object')]

# Ejercicio 56

Seleccionar filas y columnas específicas de un `DataFrame` con `iloc`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

In [ ]:
df.iloc[[1, 4], [0, 3]]

In [ ]:
type(df.iloc[[1, 4], [0, 3]])

# Ejercicio 57

Obtener la dimensión, el tamaño, y la forma de un objeto `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [23]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [24]:
df.ndim

2

In [25]:
df.size

15

In [26]:
df.shape

(5, 3)

# Ejercicio 58

Obtener la cantidad de memoria ocupada por cada columna en un `DataFrame`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [28]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = [True, False, True, False, False, True, True, False, False, True]

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos  califico
a    Oliva     11.5         1      True
b  Daniela      8.0         3     False
c     Juan     15.5         2      True
d   Germán      NaN         3     False
e   Edward      8.0         2     False
f     Alex     19.0         3      True
g    Julio     13.5         1      True
h    Edgar      NaN         1     False
i    Angie      8.0         2     False
j   Irlesa     18.0         1      True

In [29]:
df.memory_usage()

Index       80
nombre      80
puntaje     80
intentos    80
califico    10
dtype: int64

# Ejercicio 59

Seleccionar filas y columnas a partir de sus nombres con `ix` y `loc`.

## Solución:

In [30]:
import pandas as pd
import numpy as np

In [31]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [33]:
df.loc[['a', 'i'], ['nombre', 'califico']]

nombre califico
a  Oliva       Sí
i  Angie       No

In [35]:
df.iloc[[0, 8], [0, 3]]

nombre califico
a  Oliva       Sí
i  Angie       No

# Ejercicio 60

Comprobar si un objeto `DataFrame` está vacío con `empty`.

## Solución:

In [ ]:
import pandas as pd

In [36]:
df = pd.DataFrame({'X': []})

df

Empty DataFrame
Columns: [X]
Index: []

In [37]:
df.size

0

In [40]:
df.empty

True

In [41]:
import numpy as np

df = pd.DataFrame({'X': [np.nan]})

df

X
0 NaN

In [42]:
df.size

1

In [43]:
df.empty

False

# Ejercicio 61

Seleccionar los registros que tienen un campo nulo (nan).

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [44]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [45]:
df[df['puntaje'].isnull()]

nombre  puntaje  intentos califico
d  Germán      NaN         3       No
h   Edgar      NaN         1       No

# Ejercicio 62

Crear un objeto `DataFrame` a partir de objetos `Series`.

## Solución:

In [ ]:
import pandas as pd

In [ ]:
nombres = pd.Series(['Edward', 'Germán', 'Daniela'])
edades = pd.Series([27, 23, 19])

personas = {'nombre': nombres, 'edad': edades}

df = pd.DataFrame(data=personas)

df

# Ejercicio 63

Cambiar el tipo de dato de una o más columnas con `astype`.

## Solución:

In [ ]:
import pandas as pd

In [46]:
datos = {'X': [2, 3], 'Y': [5, 7]}

df = pd.DataFrame(data=datos)

df

X  Y
0  2  5
1  3  7

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   X       2 non-null      int64
 1   Y       2 non-null      int64
dtypes: int64(2)
memory usage: 160.0 bytes


In [48]:
df.dtypes

X    int64
Y    int64
dtype: object

In [55]:
df.astype({'X': 'int32'}).dtypes

X    int32
Y    int64
dtype: object

In [58]:
df.astype({'X': 'int32', 'Y': 'float64'}).dtypes

X      int32
Y    float64
dtype: object

# Ejercicio 64

Seleccionar registros donde los valores de una columna estén en un rango.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

In [ ]:
df[df['puntaje'].between(13, 19)]

# Ejercicio 65

Determinar los datos en un `DataFrame` que son nulos (NaN).

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [59]:
nombres = ['Edward', 'Daniela', 'Germán']
edades = [5, 6, np.nan]
fecha_nacimiento = [pd.Timestamp('1990-01-03'), None, np.nan]

personas = {'nombre': nombres, 'edad': edades, 'fecha_nacimiento': fecha_nacimiento}

df = pd.DataFrame(data=personas)

df

nombre  edad fecha_nacimiento
0   Edward   5.0       1990-01-03
1  Daniela   6.0              NaT
2   Germán   NaN              NaT

In [60]:
df.isna()

nombre   edad  fecha_nacimiento
0   False  False             False
1   False  False              True
2   False   True              True

# Ejercicio 66

Encontrar los datos en un `DataFrame` que **no** son nulos (NaN).

## Solución:

In [61]:
import pandas as pd
import numpy as np

In [63]:
nombres = ['Edward', 'Daniela', 'Germán']
edades = [5, 6, np.nan]
fecha_nacimiento = [pd.Timestamp('1990-01-03'), None, np.nan]

personas = {'nombre': nombres, 'edad': edades, 'fecha_nacimiento': fecha_nacimiento}

df = pd.DataFrame(data=personas)

df

nombre  edad fecha_nacimiento
0   Edward   5.0       1990-01-03
1  Daniela   6.0              NaT
2   Germán   NaN              NaT

In [64]:
df.notna()

nombre   edad  fecha_nacimiento
0    True   True              True
1    True   True             False
2    True  False             False

# Ejercicio 67

Acceder un valor o entrada específico de un objeto `DataFrame` usando `at`.

## Solución:

In [ ]:
import pandas as pd

In [65]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [66]:
df.at['b', 'Y']#batalla naval

4

In [67]:
df.at['e', 'Z'] #batalla naval

11

# Ejercicio 68

Modificar un valor o entrada específico de un objeto `DataFrame` usando `at`.

## Solución:

In [ ]:
import pandas as pd

In [ ]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

In [68]:
df.at['b', 'Y'] = 29

df

X   Y   Z
a  1   5   2
b  2  29   3
c  3   3   5
d  4   2   7
e  5   1  11

In [69]:
df.at['e', 'Z'] = 33

df

X   Y   Z
a  1   5   2
b  2  29   3
c  3   3   5
d  4   2   7
e  5   1  33

# Ejercicio 69

Modificar un valor o entrada específico de un objeto `DataFrame` usando `loc`.

## Solución:

In [ ]:
import pandas as pd

In [ ]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

In [ ]:
df.loc['a', 'Z'] = 12

df

# Ejercicio 70

Acceder a un valor o entrada específico de un objeto `DataFrame` usando `loc`.

## Solución:

In [ ]:
import pandas as pd

In [ ]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

In [ ]:
df.loc['a', 'Z']

In [ ]:
df.loc['d', 'Z']

# Ejercicio 71

Sumar los valores de una columna de un `DataFrame` con `sum`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

In [ ]:
df['intentos'].sum()

# Ejercicio 72

Acceder a una entrada de un `DataFrame` utilizando índices numéricos con `iat`.

## Solución

In [ ]:
import pandas as pd

In [ ]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

In [ ]:
df.iat[0, 2]

In [ ]:
df.iat[4, 0]

In [ ]:
df.iat[-1, 0]

In [ ]:
df.iat[-1, -1]

# Ejercicio 73

Calcular el promedio de los valores de una columna de un `DataFrame`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

In [ ]:
df['puntaje'].mean()

# Ejercicio 74

Insertar una nueva columna en una posición arbitraria con la función insert.

## Solución:

In [1]:
import pandas as pd

In [8]:
datos = {'W': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)
df

W  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [12]:
df.insert(0, '0', [0,1,2,3,4])

df

0  W  X  Y   Z
a  0  1  9  5   2
b  1  2  5  4   3
c  2  3  1  3   5
d  3  4  7  2   7
e  4  5  3  1  11

In [13]:
df['X'] = [9, 5, 1, 7, 3]

df

0  W  X  Y   Z
a  0  1  9  5   2
b  1  2  5  4   3
c  2  3  1  3   5
d  3  4  7  2   7
e  4  5  3  1  11

# Ejercicio 75

 Agregar una nueva fila a un objeto `DataFrame` por medio de `loc`.

## Solución:

In [33]:
import pandas as pd
import numpy as np

In [41]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [48]:
df.loc['k'] = ['Sebastián', 17.9, 2, 'Sí']
df

nombre  puntaje  intentos califico
a      Oliva     11.5         1       Sí
b    Daniela      8.0         3       No
c       Juan     15.5         2       Sí
d     Germán      NaN         3       No
e     Edward      8.0         2       No
f       Alex     19.0         3       Sí
g      Julio     13.5         1       Sí
h      Edgar      NaN         1       No
i      Angie      8.0         2       No
j     Irlesa     18.0         1       Sí
k  Sebastián     17.9         2       Sí

# Ejercicio 76:

Recorrer cada una de las columnas de un objeto `DataFrame` con items.

## Solución:

In [49]:
import pandas as pd

In [50]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [57]:
for columna, valor in df.items():
    print('Nombre columna:', columna)
    print('Contenido:', valor, sep= '\n')

Nombre columna: X
Contenido:
a    1
b    2
c    3
d    4
e    5
Name: X, dtype: int64
Nombre columna: Y
Contenido:
a    5
b    4
c    3
d    2
e    1
Name: Y, dtype: int64
Nombre columna: Z
Contenido:
a     2
b     3
c     5
d     7
e    11
Name: Z, dtype: int64


# Ejercicio 77

Eliminar una fila de un `DataFrame` con la función `drop`.

## Solución:

In [58]:
import pandas as pd
import numpy as np

In [59]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [62]:
df.drop('h')


nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

# Ejercicio 78

Ordenar en modo descendente un `DataFrame` por diferentes columnas.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [63]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [68]:
df.sort_values(by=['puntaje'], ascending=[False])

nombre  puntaje  intentos califico
f     Alex     19.0         3       Sí
j   Irlesa     18.0         1       Sí
c     Juan     15.5         2       Sí
g    Julio     13.5         1       Sí
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
e   Edward      8.0         2       No
i    Angie      8.0         2       No
d   Germán      NaN         3       No
h    Edgar      NaN         1       No

# Ejercicio 79

Iterar con la función `itertuples` un `DataFrame` por filas a modo de tuplas nombradas.

## Solución:

In [69]:
print('go!')

go!


In [71]:
import pandas as pd
import numpy as np

In [72]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [73]:
for registro in df.itertuples():
    print(registro)

Pandas(Index='a', X=1, Y=5, Z=2)
Pandas(Index='b', X=2, Y=4, Z=3)
Pandas(Index='c', X=3, Y=3, Z=5)
Pandas(Index='d', X=4, Y=2, Z=7)
Pandas(Index='e', X=5, Y=1, Z=11)


In [74]:
for registro in df.itertuples():
    print(registro.X, registro.Z, registro.Y * registro.X)

1 2 5
2 3 8
3 5 9
4 7 8
5 11 5


# Ejercicio 80

Reemplazar valores de una columna con la función `map` de un `DataFrame`.

## Solución:

In [75]:
import pandas as pd
import numpy as np

In [84]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [85]:
df['califico'] = df['califico'].map({'Sí': 1, 'No': 2})
#map() se utiliza para aplicar una función
# a cada elemento de una serie de pandas.
#
# En el ejemplo dado, se está utilizando para asignar
# un valor booleano a cada elemento de la columna
# 'califico' en función del valor de la cadena de texto 
# en la serie. 'Sí' se convierte en True y
# 'No' se convierte en False.

df

nombre  puntaje  intentos  califico
a    Oliva     11.5         1         1
b  Daniela      8.0         3         2
c     Juan     15.5         2         1
d   Germán      NaN         3         2
e   Edward      8.0         2         2
f     Alex     19.0         3         1
g    Julio     13.5         1         1
h    Edgar      NaN         1         2
i    Angie      8.0         2         2
j   Irlesa     18.0         1         1

# Ejercicio 81

Extraer y remover una columna por su nombre a través de la función `pop`.

## Solución:

In [ ]:
import pandas as pd

In [ ]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

In [ ]:
df.pop('Y')

In [ ]:
df

# Ejercicio 82

Reemplazar valores de una columna por medio de la función `replace`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [86]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [87]:
df['nombre'] = df['nombre'].replace('Alex', 'Alexander')

df

nombre  puntaje  intentos califico
a      Oliva     11.5         1       Sí
b    Daniela      8.0         3       No
c       Juan     15.5         2       Sí
d     Germán      NaN         3       No
e     Edward      8.0         2       No
f  Alexander     19.0         3       Sí
g      Julio     13.5         1       Sí
h      Edgar      NaN         1       No
i      Angie      8.0         2       No
j     Irlesa     18.0         1       Sí

# Ejercicio 83

Mostrar los n primeros y últimos registros de un objeto `DataFrame` con `head` y `tail`.

## Solución:

In [ ]:
import pandas as pd

In [ ]:
datos_1 = list(range(1, 101))
datos_2 = list(range(1, 101))

df = pd.DataFrame({'A': datos_1, 'B': datos_2})

df

In [ ]:
df.head()

In [ ]:
df.head(10)

In [ ]:
df.tail()

In [ ]:
df.tail(10)

# Ejercicio 84

Validar con `isin` si un elemento se halla dentro de un `DataFrame`.

## Solución:

In [88]:
import pandas as pd

In [89]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [91]:
df.isin([11, 3])

X      Y      Z
a  False  False  False
b  False  False   True
c   True   True  False
d  False  False  False
e  False  False   True

In [93]:
df.isin({'Z': [11, 3]})

X      Y      Z
a  False  False  False
b  False  False   True
c  False  False  False
d  False  False  False
e  False  False   True

# Ejercicio 85

Reemplazar valores que no cumplan una condición a través de la función `DataFrame.where`.

## Solución:

In [94]:
import pandas as pd

In [95]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [96]:
df.where(df % 2 == 0, -df) # Donde el df value analizado sea Par, reemplazalo por su (-df). 

X  Y   Z
a -1 -5   2
b  2  4  -3
c -3 -3  -5
d  4  2  -7
e -5 -1 -11

# Ejercicio 86

Agregar una nueva fila a un `DataFrame` con notación de slicing.

## Solución:

In [97]:
import pandas as pd
import numpy as np

In [109]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [112]:
df.loc['l'] = ['Pablo', 12, 1, 'Sé']
df.loc['k'] = ['Mara', 12, 1, 'Sé']
df.sort_index()

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí
k     Mara     12.0         1       Sé
l    Pablo     12.0         1       Sé

# Ejercicio 87

Usar la función `mask` para modificar los valores que cumplan una condición.

## Solución:

In [3]:
import pandas as pd

In [4]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [5]:
df.mask(df > 3) 

X    Y    Z
a  1.0  NaN  2.0
b  2.0  NaN  3.0
c  3.0  3.0  NaN
d  NaN  2.0  NaN
e  NaN  1.0  NaN

In [7]:
df.mask(df > 3, False)

X      Y      Z
a      1  False      2
b      2  False      3
c      3      3  False
d  False      2  False
e  False      1  False

# Ejercicio 88

 Iterar cada una de las filas de un `DataFrame` con la función `iterrows`.
 
 ## Solución:

In [125]:
import pandas as pd
import numpy as np

In [126]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [127]:
for indice, fila in df.iterrows():
    print(fila['nombre'], fila['califico'])

Oliva Sí
Daniela No
Juan Sí
Germán No
Edward No
Alex Sí
Julio Sí
Edgar No
Angie No
Irlesa Sí


# Ejercicio 89

Realizar consultas sobre los elementos de un objeto `DataFrame` con la función `query`.

## Solución:

In [8]:
import pandas as pd

In [9]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [10]:
df.query('X > Y')

X  Y   Z
d  4  2   7
e  5  1  11

In [11]:
df[df.X > df.Y]

X  Y   Z
d  4  2   7
e  5  1  11

# Ejercicio 90

Realizar operaciones aritméticas básicas entre dos objetos `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [12]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [13]:
df.add(df)

X   Y   Z
a   2  10   4
b   4   8   6
c   6   6  10
d   8   4  14
e  10   2  22

In [14]:
df.sub(df)

X  Y  Z
a  0  0  0
b  0  0  0
c  0  0  0
d  0  0  0
e  0  0  0

In [15]:
df.mul(df)

X   Y    Z
a   1  25    4
b   4  16    9
c   9   9   25
d  16   4   49
e  25   1  121

In [16]:
df.div(df)

X    Y    Z
a  1.0  1.0  1.0
b  1.0  1.0  1.0
c  1.0  1.0  1.0
d  1.0  1.0  1.0
e  1.0  1.0  1.0

In [17]:
datos = {'X': [5, 4, 3, 2, 1], 'Y': [1, 2, 3, 4, 5], 'Z': [2, 3, 5, 7, 11]}

df2 = pd.DataFrame(datos, index=indices)

df2

X  Y   Z
a  5  1   2
b  4  2   3
c  3  3   5
d  2  4   7
e  1  5  11

In [18]:
df.add(df2)

X  Y   Z
a  6  6   4
b  6  6   6
c  6  6  10
d  6  6  14
e  6  6  22

In [25]:
df**df2

X   Y             Z
a   1   5             4
b  16  16            27
c  27  27          3125
d  16  16        823543
e   5   1  285311670611

In [19]:
df.sub(df2)

X  Y  Z
a -4  4  0
b -2  2  0
c  0  0  0
d  2 -2  0
e  4 -4  0

In [20]:
df.mul(df2)

X  Y    Z
a  5  5    4
b  8  8    9
c  9  9   25
d  8  8   49
e  5  5  121

In [21]:
df.div(df2)

X    Y    Z
a  0.2  5.0  1.0
b  0.5  2.0  1.0
c  1.0  1.0  1.0
d  2.0  0.5  1.0
e  5.0  0.2  1.0

# Ejercicio 91

Obtener como una lista Python los nombres de las columnas de un `DataFrame`.

## Solución:

In [26]:
import pandas as pd

In [27]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [28]:
type(df.columns.values)

numpy.ndarray

In [29]:
list(df.columns.values)

['X', 'Y', 'Z']

# Ejercicio 92

Calcular el módulo sobre cada elemento de un objeto `DataFrame`.

## Solución:

In [31]:
import pandas as pd

In [30]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [32]:
df.mod(2)

X  Y  Z
a  1  1  0
b  0  0  1
c  1  1  1
d  0  0  1
e  1  1  1

In [33]:
df.mod(5)

X  Y  Z
a  1  0  2
b  2  4  3
c  3  3  0
d  4  2  2
e  0  1  1

# Ejercicio 93

Renombrar las columnas de un objeto `DataFrame`.

## Solución:

In [34]:
import pandas as pd

In [35]:
datos = {'Columna1': [1, 2, 3, 4, 5], 'Columna2': [5, 4, 3, 2, 1], 
         'Columna3': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

Columna1  Columna2  Columna3
a         1         5         2
b         2         4         3
c         3         3         5
d         4         2         7
e         5         1        11

In [36]:
renombrado_columnas = {'Columna1': 'X', 'Columna2': 'Y', 'Columna3': 'Z'}

df.rename(columns=renombrado_columnas)

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

# Ejercicio 94

Calcular el producto punto entre dos objetos `DataFrame` con la función `dot`.

## Solución:

In [37]:
import pandas as pd

In [38]:
datos_1 = [[1, 2, 3, 4, 5], [5, 4, 3, 2, 1], [2, 3, 5, 7, 11]]

datos_2 = [[1, 2, 3], [9, 8, 7], [1, 1, 1], [3, 2, 1], [8, 5, 2]]

df_1 = pd.DataFrame(datos_1)
df_2= pd.DataFrame(datos_2)

In [39]:
df_1

0  1  2  3   4
0  1  2  3  4   5
1  5  4  3  2   1
2  2  3  5  7  11

In [40]:
df_2

0  1  2
0  1  2  3
1  9  8  7
2  1  1  1
3  3  2  1
4  8  5  2

In [41]:
df_1.dot(df_2)

0    1   2
0   74   54  34
1   58   54  50
2  143  102  61

# Ejercicio 95

Seleccionar registros donde se cumpla una condición para valores de una columna.

## Solución:

In [43]:
import pandas as pd
import numpy as np

In [44]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [51]:
df.loc[df['califico'] == 'Sí']

nombre  puntaje  intentos califico
a   Oliva     11.5         1       Sí
c    Juan     15.5         2       Sí
f    Alex     19.0         3       Sí
g   Julio     13.5         1       Sí
j  Irlesa     18.0         1       Sí

# Ejercicio 96

Usar operadores relacionales para comparar elementos de objetos `DataFrame`.

## Solución:

In [52]:
import pandas as pd

In [53]:
datos_1 = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

datos_2 = {'X': [8, 9, 3, 2, 1], 'Y': [9, 2, 0, 1, 1], 'Z': [2, 4, 13, 2, 11]}

df_1 = pd.DataFrame(data=datos_1)
df_2 = pd.DataFrame(data=datos_2)

In [ ]:
df_1

In [ ]:
df_2

In [54]:
df_1 < df_2

X      Y      Z
0   True   True  False
1   True  False   True
2  False  False   True
3  False  False  False
4  False  False  False

In [ ]:
df_1 > df_2

In [ ]:
df_1 <= df_2

In [ ]:
df_1 >= df_2

In [55]:
df_1 == df_2

X      Y      Z
0  False  False   True
1  False  False  False
2   True  False  False
3  False  False  False
4  False   True   True

# Ejercicio 97

Usar métodos de instancia de `DataFrame` para comparacionese de igualdad.

## Solución:

In [ ]:
import pandas as pd

In [58]:
datos_1 = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

datos_2 = {'X': [8, 9, 3, 2, 1], 'Y': [9, 2, 0, 1, 1], 'Z': [2, 4, 13, 2, 11]}

df_1 = pd.DataFrame(data=datos_1)
df_2 = pd.DataFrame(data=datos_2)

In [ ]:
df_1

In [ ]:
df_2

In [59]:
df_1.eq(df_2)

X      Y      Z
0  False  False   True
1  False  False  False
2   True  False  False
3  False  False  False
4  False   True   True

In [60]:
df_1.ne(df_2)

X      Y      Z
0   True   True  False
1   True   True   True
2  False   True   True
3   True   True   True
4   True  False  False

# Ejercicio 98

Cambiar el orden de dos columnas de un objeto `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [61]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [62]:
df = df[['X', 'Z', 'Y']]

df

X   Z  Y
a  1   2  5
b  2   3  4
c  3   5  3
d  4   7  2
e  5  11  1

# Ejercicio 99

Combinar dos `DataFrame` y dejar en cada columna el valor mínimo de la suma por columna.

## Solución:

In [63]:
import pandas as pd

In [64]:
datos_1 = {'A': [1, 1], 'B': [5, 5]}
datos_2 = {'A': [3, 3], 'B': [2, 2]}

df_1 = pd.DataFrame(data=datos_1)
df_2 = pd.DataFrame(data=datos_2)

In [65]:
df_1

A  B
0  1  5
1  1  5

In [66]:
df_2

A  B
0  3  2
1  3  2

In [67]:
minimo_columna = lambda s1, s2: s1.sum() if s1.sum() < s2.sum() else s2.sum()

df_1.combine(df_2, minimo_columna)

A  B
0  2  4
1  2  4

# Ejercicio 100

Agregar una nueva fila a una `DataFrame` con el método `append`.

## Solución:

In [ ]:
import pandas as pd

In [68]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [70]:
nuevo_registro = {'X': 6, 'Y': 0, 'Z': 13}

df = df.append(nuevo_registro, ignore_index=True)

df

C:\Users\Luca\AppData\Local\Temp\ipykernel_14440\2729266879.py:3: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(nuevo_registro, ignore_index=True)


X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11
5  6  0  13
6  6  0  13

# Ejercicio 101

Aplicar una función a todos los elementos de un `DataFrame` con la función `apply`.

## Solución:

In [71]:
import pandas as pd

In [72]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [73]:
import numpy as np

df.apply(np.sqrt) #la raíz cuadrada positiva

X         Y         Z
0  1.000000  2.236068  1.414214
1  1.414214  2.000000  1.732051
2  1.732051  1.732051  2.236068
3  2.000000  1.414214  2.645751
4  2.236068  1.000000  3.316625

In [74]:
def cuadrado(x):
    return x * x

df.apply(cuadrado)

X   Y    Z
0   1  25    4
1   4  16    9
2   9   9   25
3  16   4   49
4  25   1  121

In [75]:
df.apply(np.sum, axis=0)

X    15
Y    15
Z    28
dtype: int64

In [76]:
df.apply(np.sum, axis=1)

0     8
1     9
2    11
3    13
4    17
dtype: int64

In [77]:
df.apply(lambda x: x * x)

X   Y    Z
0   1  25    4
1   4  16    9
2   9   9   25
3  16   4   49
4  25   1  121

# Ejercicio 102

Guardar los registros de un `DataFrame` en un archivo de formato **CSV**.

## Solución:

In [78]:
import pandas as pd
import numpy as np

In [79]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [80]:
df.to_csv('jugadores.csv', index=False)